samples/bluetooth/direct_test_mode/src/fem/generic_fem.c

```diff
@@ -0,0 +1,587 @@
+#elif defined(CONFIG_HAS_HW_NRF_DPPIC)
+#include <nrfx_dppi.h>
+#define gppi_channel_t uint8_t
+#define gppi_channel_alloc nrfx_dppi_channel_alloc
+#else
+#error "No PPI or DPPI"
+#endif
+
+#define GENERIC_FEM_TIMER_INSTANCE 2
+
+#define GENERIC_FEM_NODE DT_NODELABEL(nrf_radio_fem)
+
+#define CTX_REG DT_REG_ADDR(DT_PHANDLE(GENERIC_FEM_NODE, ctx_gpios))
+#define CTX_PIN DT_GPIO_PIN(GENERIC_FEM_NODE, ctx_gpios)
+#define CTX_FLAGS DT_GPIO_FLAGS(GENERIC_FEM_NODE, ctx_gpios)
+
+#define CRX_REG DT_REG_ADDR(DT_PHANDLE(GENERIC_FEM_NODE, crx_gpios))
+#define CRX_PIN DT_GPIO_PIN(GENERIC_FEM_NODE, crx_gpios)
+#define CRX_FLAGS DT_GPIO_FLAGS(GENERIC_FEM_NODE, crx_gpios)
+
+#define CTX_SETTLE_TIME DT_PROP(GENERIC_FEM_NODE, ctx_settle_time_us)
+#define CRX_SETTLE_TIME DT_PROP(GENERIC_FEM_NODE, crx_settle_time_us)
+
+#define GENERIC_FEM_GPIO_POLARITY_GET(_dt_property)                   \
+	((DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), _dt_property) & \
+	GPIO_ACTIVE_LOW) ? false : true)
+
+struct gpiote_pin {
+	uint32_t abs_pin;
+	uint8_t gpiote_channel;
+	bool active_high;
+};
+
+static struct generic_fem {
+	struct gpiote_pin ctx_pin;
+	struct gpiote_pin crx_pin;
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), ant_sel_gpios)
+	const struct gpio_dt_spec ant_sel;
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), csd_gpios)
+	const struct gpio_dt_spec csd;
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), cps_gpios)
+	const struct gpio_dt_spec cps;
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), chl_gpios)
+	const struct gpio_dt_spec chl;
+#endif
+	gppi_channel_t timer_ch;
+	gppi_channel_t pin_set_ch;
+	gppi_channel_t pin_clr_ch;
+	const nrfx_timer_t timer;
+} generic_fem_cfg = {
+	.timer = NRFX_TIMER_INSTANCE(GENERIC_FEM_TIMER_INSTANCE),
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), ant_sel_gpios)
+	.ant_sel = GPIO_DT_SPEC_GET(GENERIC_FEM_NODE, ant_sel_gpios),
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), csd_gpios)
+	.csd = GPIO_DT_SPEC_GET(GENERIC_FEM_NODE, csd_gpios),
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), cps_gpios)
+	.cps = GPIO_DT_SPEC_GET(GENERIC_FEM_NODE, cps_gpios),
+#endif
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), chl_gpios)
+	.chl = GPIO_DT_SPEC_GET(GENERIC_FEM_NODE, chl_gpios)
+#endif
+};
+
+enum generic_fem_trx {
+	GENERIC_FEM_TX,
+	GENERIC_FEM_RX
+};
+
+/* Generic Front-end module (FEM) state */
+enum fem_state {
+	/* Front-end module Power-Down state */
+	FEM_STATE_OFF,
+
+	/* Front-end module is powered and ready for PA/LNA operation
+	 * (program state).
+	 */
+	FEM_STATE_READY,
+
+	/* Front-end module is in PA/LNA state or processes command. */
+	FEM_STATE_BUSY
+};
+
+static uint32_t active_evt;
+static uint32_t deactive_evt;
+
+static atomic_t state = ATOMIC_INIT(FEM_STATE_READY);
+
+static uint32_t gpio_port_num_decode(NRF_GPIO_Type *reg)
+{
+	if (reg == NRF_P0) {
+		return 0;
+	}
+#if CONFIG_HAS_HW_NRF_GPIO1
+	else if (reg == NRF_P1) {
```

Comment:
consider using switch statement (with default case routing to ASSERT)

Reply:
This a pointer comparison so switch case cannot be used here. It would be possible after casting but I think that it would be more complicated than just if statement